design/one-pager-k8s-native-providers.md

```diff
@@ -0,0 +1,287 @@
+
+While the `KubernetesApplication` pattern detailed above provides a starting
+point for this design, a key difference is that `KubernetesApplication` does not
+adhere to the separation of concern structure of other Crossplane resources
+because Kubernetes resources are not modeled in a `claim <-> portable class <->
+non-portable class <-> managed <-> external` workflow. Kubernetes-native
+infrastructure resources, on the other hand, *should* be modeled with the same
+separation of concern that other cloud-provider resources expose.
+
+## Workflow Design
+
+The general design from a user experience perspective should differ very little
+from utilizing any other infrastructure provider stack. Developers should be
+able to request the creation of an abstract resource type via a claim, and
+cluster operators should be able to define the classes of service available to
+satisfy that claim kind.
+
+As an example, Rook provides the option to provision [Minio] object storage in a
+Kubernetes cluster. Assuming you have installed Rook and the corresponding Minio
+operator into a cluster, you should be able to dynamically provision a Minio
+object storage cluster into the Kubernetes cluster from your Crossplane control
+cluster. The experience would look as follows:
+
+1. Create a Kubernetes `Provider` for target cluster:
+
+```yaml
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: demo-provider-rook
+  namespace: rook-infra-dev
+type: Opaque
+data:
+  endpoint: MY_K8S_ENDPOINT
+  username: MY_K8S_USERNAME
+  password: MY_K8S_PASSWORD
+  clusterCA: MY_K8S_CLUSTER_CA
+  clientCert: MY_K8S_CLIENT_CERT
+  clientKey: MY_K8S_CLIENT_KEY
+  token: MY_K8S_TOKEN
+---
+apiVersion: kubernetes.crossplane.io/v1alpha2
+kind: Provider
+metadata:
+  name: demo-kubernetes
+  namespace: azure-infra-dev
+spec:
+  credentialsSecretRef:
+    name: demo-provider-rook
+```
+
+2. Create a non-portable Rook Minio `ObjectStoreClass`
+
+```yaml
+apiVersion: minio.rook.crossplane.io/v1alpha1
+kind: ObjectStoreClass
+metadata:
+  name: rook-minio
+  namespace: rook-infra-dev
+spec:
+  scope:
+    nodeCount: 4
+    volumeClaimTemplates:
+    - metadata:
+        name: rook-minio-data1
+      spec:
+        accessModes: [ "ReadWriteOnce" ]
+        resources:
+          requests:
+            storage: "8Gi"
+  # A key value list of annotations
+  annotations:
+  #  key: value
+  placement:
+    tolerations:
+    nodeAffinity:
+    podAffinity:
+    podAnyAffinity:
+  credentials:
+    name: minio-my-store-access-keys
+    namespace: rook-minio
+  providerRef:
+    name: demo-kubernetes
+    namespace: rook-infra-dev
+```
+
+3. Create a portable `BucketClass`
+
+```yaml
+apiVersion: storage.crossplane.io/v1alpha1
+kind: BucketClass
+metadata:
+  name: bucket-standard
+  namespace: app-project1-dev
+classRef:
+  kind: ObjectStoreClass
+  apiVersion: minio.rook.crossplane.io/v1alpha1
+  name: rook-minio
+  namespace: rook-infra-dev
+```
+
+4. Create a `Bucket` claim
```

Comment:
so an important distinction to make here is that the `ObjectStoreClass` that this `Bucket` claim will use via `Bucket` -> `BucketClass` -> `ObjectStoreClass` is actually the entire underlying object storage solution, not just a single bucket.  The analogy here for AWS is the entire S3 service.  Furthermore, when you create a Minio object store, it doesn't start with any buckets at all.  As written, this claim -> class relationship would result in the user requesting a bucket but instead getting a blank object store with 0 buckets.

We need to think through a bit more on how the user requesting a Bucket can result in a bucket getting created for them, and how to deal with the cases where the Minio object store (or any of the other object storage providers in Rook) does not exist yet and already exists.

Steps to create a Minio object store:
1. Deploy the Minio operator
1. Create an instance of an objectstore.minio.rook.io
1. Create a bucket (there is currently no CRD for this) in that object store

Reply:
Thanks for catching this @jbw976 - it seems like it could expand the scope of this work, at least a little. A bunch of questions come to mind

Am I correct in my understanding that this concern only applies to Rook's object storage types, and not the SQL "instance" types? It seems like deploying CockroachDB or Yugabyte with Rook gets you similar functionality to deploying a CloudSQL instance; i.e. a Postgres-compliant API you could use to create and manage databases. So it seems to make sense that a deployment of CockroachDB, for example, would satisfy a Crossplane `PostgreSQLInstance` claim.

Is there some way we could reuse or extend the `S3Bucket.storage.aws.crossplane.io` managed resource for use with S3 compatible object stores? Perhaps it would be better to duplicate it as `S3Bucket.storage.rook.crossplane.io` alongside the existing, especially given folks may not want to install `stack-aws` just to be able to use buckets with `stack-rook`?

How would deployment of an object store API (e.g. Minio) fit into Crossplane's model? Given Rook provides us with multiple ways to deploy an S3-compatible object store would it make sense to add an `ObjectStore` resource claim type that could eventually be satisfied by Minio, EdgeFS, or Ceph managed resources?

What would the user experience be, assuming we were had already deployed Rook (but no storage providers) and configured a Kubernetes provider appropriately? Something like:

1. Create a `MinioClass`, point an `ObjectStoreClass` at it.
1. Create a `ObjectStore` claim, point it at the `ObjectStoreClass` from step 1.

Then, for each bucket we wanted to create:
1. Create a `S3BucketClass` non-portable class, point a `BucketClass` at it.
1. Create a `Bucket` claim, point it at the `BucketClass` from step 1.

Would we need to configure some kind of `provider.s3.rook.crossplane.io` to actually connect to and use the API exposed by the Rook object storage providers? Would we need to explicitly create a user in order to get credentials to add that provider?

Reply:
@jbw976 thanks for catching this, I was much too hasty in translating the doc from CockroachDB to Minio, so I appreciate your watchful eye! I broadly agree with @negz thoughts above, although I am hesitant to add a generic claim type that is only used by one provider. It seems that it may be an opportunity to introduce a "provider-specific" claim for `ObjectStore`. However, maybe the generic `ObjectStore` claim does make sense because Rook is basically just an API for many other in-cluster storage providers.